rasa/core/domain.py

```diff
@@ -259,6 +259,8 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
         # it is super important to sort the slots here!!!
         # otherwise state ordering is not consistent
         slots = []
+        # make a copy to not alter the input dictionary
```

Comment:
what are pros / cons in selectively copying parts of the domain dictionary vs. doing deepcopy of the whole domain at the start of `from_dict`?

Reply:
Pro for doing it selectively:
The two functions this affects both start with `collect_`. That seems to imply that they only gather stuff without changing anything. Making a copy in each of them guarantees that no unexpected side-effects show up if they are ever used elsewhere (e.g. `collect_intent_properties` is used in the constructor for `Domain`, so you could imagine this being called via there instead of via `from_dict`).

Pro for doing it once in `from_dict`:
More obvious / easy. Also guarantees that if the `from_dict` function ever changes we can't accidentally regress to altering the dictionary again somewhere else in it.

How about doing it both? That is a bit redundant, but ensures things stay safe in the future.

Reply:
One more `pro` for your current approach is that it's clear why `copy` is required since it's right next to the code which alters the input. I think doing it both doesn't make sense because we get all the cons and not all the pros (if that makes sense 😀). I am slightly (like 55% to 45%) leaning towards keeping it the way it currently is.